### ISSUE TITLE
traefik crashes when kubernetes secret is used for basic auth

### ISSUE THREAD

**Issue by Scukerman** [containous/traefik#1794]
### Do you want to request a *feature* or report a *bug*?
BUG.

### What did you do?
Applied #1707 to get access to secrets
Installed deployment of latest traefik image with 3 replicas.
Deployed k8s-dashboard and made ingress rule with basic auth + secret.

### What did you expect to see?

All replicas to be working. Not only 2 of them.

### What did you see instead?

2\3 are up and the third one couldn't get existing secret.

### Output of `traefik version`: (_What version of Traefik are you using?_)
```
Version:      v1.3.1
Codename:     raclette
Go version:   go1.8.3
Built:        2017-06-16_11:21:48AM
OS/Arch:      linux/amd64
```
### What is your environment & configuration (arguments, toml, provider, platform, ...)?

```toml
debug = true
logLevel = "INFO"
defaultEntryPoints = ["http","https"]

[entryPoints]
	[entryPoints.http]
	address = ":80"
	[entryPoints.https]
	address = ":443"

[web]
address = ":8080"
	[web.auth.basic]
	users = [ "test:$apr1$Y8X0T234$8hoaJeKIka4wJi13yETbK." ]

[kubernetes]
namespaces = [ "default", "kube-system" ]
```

### If applicable, please paste the log output in debug mode (`--debug` switch)

```
time="2017-06-27T15:32:58Z" level=info msg="Traefik version v1.3.1 built on 2017-06-16_11:21:48AM" 
time="2017-06-27T15:32:58Z" level=info msg="Using TOML configuration file /config/traefik.toml" 
time="2017-06-27T15:32:58Z" level=debug msg="Global configuration loaded {"GraceTimeOut":10000000000,"Debug":true,"CheckNewVersion":true,"AccessLogsFile":"","TraefikLogsFile":"","LogLevel":"DEBUG","EntryPoints":{"http":{"Network":"","Address":":80","TLS":null,"Redirect":null,"Auth":null,"Compress":false},"https":{"Network":"","Address":":443","TLS":null,"Redirect":null,"Auth":null,"Compress":false}},"Cluster":null,"Constraints":[],"ACME":null,"DefaultEntryPoints":["http","https"],"ProvidersThrottleDuration":2000000000,"MaxIdleConnsPerHost":200,"IdleTimeout":180000000000,"InsecureSkipVerify":false,"Retry":null,"HealthCheck":{"Interval":30000000000},"Docker":null,"File":null,"Web":{"Address":":8080","CertFile":"","KeyFile":"","ReadOnly":false,"Statistics":null,"Metrics":null,"Path":"","Auth":{"Basic":{"Users":["test:$apr1$Y8X0T234$8hoaJeKIka4wJi13yETbK."],"UsersFile":""},"Digest":null,"HeaderField":""}},"Marathon":null,"Consul":null,"ConsulCatalog":null,"Etcd":null,"Zookeeper":null,"Boltdb":null,"Kubernetes":{"Watch":true,"Filename":"","Constraints":[],"Endpoint":"","Token":"","CertAuthFilePath":"","DisablePassHostHeaders":false,"Namespaces":["default","kube-system"],"LabelSelector":""},"Mesos":null,"Eureka":null,"ECS":null,"Rancher":null,"DynamoDB":null}" 
time="2017-06-27T15:32:58Z" level=info msg="Preparing server https &{Network: Address::443 TLS:<nil> Redirect:<nil> Auth:<nil> Compress:false}" 
time="2017-06-27T15:32:58Z" level=info msg="Preparing server http &{Network: Address::80 TLS:<nil> Redirect:<nil> Auth:<nil> Compress:false}" 
time="2017-06-27T15:32:58Z" level=info msg="Starting provider *server.WebProvider {"Address":":8080","CertFile":"","KeyFile":"","ReadOnly":false,"Statistics":null,"Metrics":null,"Path":"","Auth":{"Basic":{"Users":["test:$apr1$Y8X0T234$8hoaJeKIka4wJi13yETbK."],"UsersFile":""},"Digest":null,"HeaderField":""}}" 
time="2017-06-27T15:32:58Z" level=info msg="Starting provider *kubernetes.Provider {"Watch":true,"Filename":"","Constraints":[],"Endpoint":"","Token":"","CertAuthFilePath":"","DisablePassHostHeaders":false,"Namespaces":["default","kube-system"],"LabelSelector":""}" 
time="2017-06-27T15:32:58Z" level=info msg="Starting server on :80" 
time="2017-06-27T15:32:58Z" level=info msg="Starting server on :443" 
time="2017-06-27T15:32:58Z" level=info msg="Creating in-cluster Provider client
" 
time="2017-06-27T15:32:58Z" level=debug msg="Using label selector: ''" 
time="2017-06-27T15:32:59Z" level=debug msg="Received event from kubernetes &Endpoints{ObjectMeta:ObjectMeta{Name:tiller-deploy,GenerateName:,Namespace:kube-system,SelfLink:/api/v1/namespaces/kube-system/endpoints/tiller-deploy,UID:db0c7d32-1b98-11e7-8d60-0025905d1042,ResourceVersion:9558297,Generation:0,CreationTimestamp:2017-04-07 13:48:18 +0000 UTC,DeletionTimestamp:<nil>,DeletionGracePeriodSeconds:nil,Labels:map[string]string{app: helm,name: tiller,},Annotations:map[string]string{},OwnerReferences:[],Finalizers:[],ClusterName:,},Subsets:[{[{192.168.0.4  0xc4207bcbc0 ObjectReference{Kind:Pod,Namespace:kube-system,Name:tiller-deploy-1332042700-pb18c,UID:daebcdb3-1b98-11e7-8d60-0025905d1042,APIVersion:,ResourceVersion:9558296,FieldPath:,}}] [] [{tiller 44134 TCP}]}],}" 
time="2017-06-27T15:32:59Z" level=error msg="Provider connection error secret "kube-system"/"k8s-dashboard-basic-auth" not found, retrying in 651.281427ms" 
ERROR: logging before flag.Parse: E0627 15:32:59.062442       1 runtime.go:66] Observed a panic: "send on closed channel" (send on closed channel)
/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/pkg/util/runtime/runtime.go:72
/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/pkg/util/runtime/runtime.go:65
/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/pkg/util/runtime/runtime.go:51
/usr/local/go/src/runtime/asm_amd64.s:514
/usr/local/go/src/runtime/panic.go:489
/usr/local/go/src/runtime/chan.go:180
/usr/local/go/src/runtime/chan.go:604
/go/src/github.com/containous/traefik/provider/kubernetes/client.go:136
/go/src/github.com/containous/traefik/provider/kubernetes/client.go:143
/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/tools/cache/controller.go:173
<autogenerated>:50
/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/tools/cache/controller.go:251
/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/tools/cache/delta_fifo.go:420
/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/tools/cache/controller.go:131
/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/tools/cache/controller.go:102
/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/pkg/util/wait/wait.go:96
/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/pkg/util/wait/wait.go:97
/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/pkg/util/wait/wait.go:52
/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/tools/cache/controller.go:102
/usr/local/go/src/runtime/asm_amd64.s:2197
ERROR: logging before flag.Parse: E0627 15:32:59.062615       1 runtime.go:66] Observed a panic: "send on closed channel" (send on closed channel)
/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/pkg/util/runtime/runtime.go:72
/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/pkg/util/runtime/runtime.go:65
/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/pkg/util/runtime/runtime.go:51
/usr/local/go/src/runtime/asm_amd64.s:514
/usr/local/go/src/runtime/panic.go:489
/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/pkg/util/runtime/runtime.go:58
/usr/local/go/src/runtime/asm_amd64.s:514
/usr/local/go/src/runtime/panic.go:489
/usr/local/go/src/runtime/chan.go:180
/usr/local/go/src/runtime/chan.go:604
/go/src/github.com/containous/traefik/provider/kubernetes/client.go:136
/go/src/github.com/containous/traefik/provider/kubernetes/client.go:143
/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/tools/cache/controller.go:173
<autogenerated>:50
/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/tools/cache/controller.go:251
/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/tools/cache/delta_fifo.go:420
/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/tools/cache/controller.go:131
/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/tools/cache/controller.go:102
/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/pkg/util/wait/wait.go:96
/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/pkg/util/wait/wait.go:97
/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/pkg/util/wait/wait.go:52
/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/tools/cache/controller.go:102
/usr/local/go/src/runtime/asm_amd64.s:2197
panic: send on closed channel [recovered]
	panic: send on closed channel [recovered]
	panic: send on closed channel

goroutine 39 [running]:
github.com/containous/traefik/vendor/k8s.io/client-go/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/pkg/util/runtime/runtime.go:58 +0x126
panic(0x1ca6940, 0xc4208ee540)
	/usr/local/go/src/runtime/panic.go:489 +0x2cf
github.com/containous/traefik/vendor/k8s.io/client-go/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/pkg/util/runtime/runtime.go:58 +0x126
panic(0x1ca6940, 0xc4208ee540)
	/usr/local/go/src/runtime/panic.go:489 +0x2cf
github.com/containous/traefik/provider/kubernetes.eventHandlerFunc(0xc420112d20, 0x1f5e6e0, 0xc4200f3800)
	/go/src/github.com/containous/traefik/provider/kubernetes/client.go:136 +0x55
github.com/containous/traefik/provider/kubernetes.newResourceEventHandlerFuncs.func1(0x1f5e6e0, 0xc4200f3800)
	/go/src/github.com/containous/traefik/provider/kubernetes/client.go:143 +0x3e
github.com/containous/traefik/vendor/k8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnAdd(0xc42032dc50, 0xc42032dc60, 0xc42032dc70, 0x1f5e6e0, 0xc4200f3800)
	/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/tools/cache/controller.go:173 +0x49
github.com/containous/traefik/vendor/k8s.io/client-go/tools/cache.(*ResourceEventHandlerFuncs).OnAdd(0xc4207ce300, 0x1f5e6e0, 0xc4200f3800)
	<autogenerated>:50 +0x73
github.com/containous/traefik/vendor/k8s.io/client-go/tools/cache.NewInformer.func1(0x1cd5ac0, 0xc4208ff2e0, 0x1cd5ac0, 0xc4208ff2e0)
	/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/tools/cache/controller.go:251 +0x24d
github.com/containous/traefik/vendor/k8s.io/client-go/tools/cache.(*DeltaFIFO).Pop(0xc4206602c0, 0xc4205bf320, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/tools/cache/delta_fifo.go:420 +0x218
github.com/containous/traefik/vendor/k8s.io/client-go/tools/cache.(*Controller).processLoop(0xc42015f110)
	/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/tools/cache/controller.go:131 +0x3c
github.com/containous/traefik/vendor/k8s.io/client-go/tools/cache.(*Controller).(github.com/containous/traefik/vendor/k8s.io/client-go/tools/cache.processLoop)-fm()
	/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/tools/cache/controller.go:102 +0x2a
github.com/containous/traefik/vendor/k8s.io/client-go/pkg/util/wait.JitterUntil.func1(0xc4207d87b0)
	/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/pkg/util/wait/wait.go:96 +0x5e
github.com/containous/traefik/vendor/k8s.io/client-go/pkg/util/wait.JitterUntil(0xc4207a7fb0, 0x3b9aca00, 0x0, 0x1b63001, 0xc420112c60)
	/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/pkg/util/wait/wait.go:97 +0xad
github.com/containous/traefik/vendor/k8s.io/client-go/pkg/util/wait.Until(0xc4207d87b0, 0x3b9aca00, 0xc420112c60)
	/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/pkg/util/wait/wait.go:52 +0x4d
github.com/containous/traefik/vendor/k8s.io/client-go/tools/cache.(*Controller).Run(0xc42015f110, 0xc420112c60)
	/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/tools/cache/controller.go:102 +0x1c7
created by github.com/containous/traefik/provider/kubernetes.(*clientImpl).WatchSecrets
	/go/src/github.com/containous/traefik/provider/kubernetes/client.go:227 +0x28e
```

**Comment by timoreimann** [comment]
The log says that we are sending on a closed channel, referencing [this piece of the code](https://github.com/containous/traefik/blob/2201dcd50507702ad881e60e8d633c809ddf5df1/provider/kubernetes/client.go#L136) where we forward an object received via the Informer to the events channel.

I think the problem is due to a race between our `WatchAll()` method closing all channels once the goroutine returns on reception of the stop signal, and the Informer-internal processing trying to notify our `eventHandlerFunc`.

**Comment by timoreimann** [comment]
Note to myself and the maintainers: Our naming of the various involved channels (`stop`, `watch`, `event`) is super confusing and continuously changing depending on what method you're looking at. :/

**Comment by timoreimann** [comment]
I'll be taking a wild shot here: could the problem be due to us not waiting for the secrets controller to have synchronized [here](https://github.com/containous/traefik/blob/2201dcd50507702ad881e60e8d633c809ddf5df1/provider/kubernetes/client.go#L266)?

/cc @dtomcej @errm

**Comment by timoreimann** [comment]
I am able to reproduce this by setting up basic auth via annotations but don't provide a Secret object.

**Comment by timoreimann** [comment]
Yep, waiting for the secret controller sync fixes things.

PR incoming...

**Comment by nmengin** [comment]
Closed by #1798 .
